### DISCUSSION TITLE
Generative Replay Approach

### DISCUSSION THREAD

**Discussion by Alparslan-Tamer** [ContinualAI/avalanche#414] [Feature Request]
Firstly, thank you for this framework, this is a lot help to me.
Is it possible for you to add the generative replay approach to this library (any of them)?
thanks in advance

**Comment by AndreaCossu** [comment]
Hi @Alparslan-Tamer! Generative replay approaches are definitely compatible with Avalanche. I'll add it as a feature request. Thank you :smile:

**Comment by ggraffieti** [comment]
Hi @Alparslan-Tamer, 
you can use the "from scratch mode" for the strategy creation ([see here](https://avalanche.continualai.org/from-zero-to-hero-tutorial/3.-training#create-your-strategy)), in this case you are totally free regarding the CL strategy to use, just insert your generative replay code inside the `train` method, that's all! In any case you have access to any other Avalanche functionality, such as predefined benchmarks. 
Another possibility is to use the "callbacks mode" ([see here)](https://avalanche.continualai.org/from-zero-to-hero-tutorial/3.-training#callbacks-mode). In this case the code for training the generator can be inserted in the `before_training_exp` callback (or `after_training_exp`, depending on when you want to train your generative model) and the sampling could be done inside `adapt_train_dataset` callback. 

In any case we are planning a basic generative replay plugin, if you want you can try to implement one by yourself and make a contribution to Avalanche! :tada:

**Reply by AntonioCarta** [reply]
I just want to add that we already have an implementation of generative replay on colab [here](https://github.com/ContinualAI/colab/blob/master/notebooks/intro_to_generative_replay.ipynb). You can start from that and hopefully we will have it on avalanche ASAP.

**Comment by Alparslan-Tamer** [comment]
Hi guys, 
thanks for the answers. I will try to develop the generative replay code you threw (with avalanche). But I'm new in this job so I'm not sure if I can. But if I do, I will share with you the new version (a few varieties).
Also sorry for my bad english.

**Comment by moro-no-kimi** [comment]
This may be of some assistance: https://github.com/TLESORT/Generative_Continual_Learning

**Comment by travela** [comment]
Hi Avalanche team! :tada: 
I am currently working on an implementation of a Generative Replay Plugin / Template (as part of my master thesis on Continual Learning and Generative Replay in particular). 
I would love to contribute to the project. I've been watching it for a while now and I am familiar with the codebase +  Contributing Guidelines. 

Therefore, I would like to propose that I create an issue from this and assign it to me. Let me know what you think and if there's anything else I should be aware of!  :smile:

**Reply by AntonioCarta** [reply]
Great @travela ! This is something that we wanted to add ourselves. You can open an issue and assign it to yourself. If you need any help you can also ping me directly on the ContinualAI slack.

**Reply by travela** [reply]
Awesome! Thanks @AntonioCarta, I will probably ping you soon :)
